### PR TITLE
Addon disabled listener

### LINKED ISSUE #1303: Ability for addons to know they've been disabled
Possible usecases:
1. Userscripts could revert their action in the page if disabled (e.g. discuss button is removed after it is disabled)
2. Could be used in persistent scripts as the proper way to unregister any event listeners.

### LINKED ISSUE #1495: Make the pen extension be a main categorey
**Is your feature request related to a problem? Please describe.**
I don't like how pen is now an extension when it used to be a main categorey

**Describe the solution you'd like**
The pen extension is auto added

**Describe alternatives you've considered**
No alternitives.

**Additional context**
This would be awesome

### LINKED ISSUE #1670: Enabling or disabling addons in the popup should not cause them to move around
See https://github.com/ScratchAddons/ScratchAddons/issues/1670#issuecomment-826062306
Original description:

**Describe the bug**

Disabling themes with no running userscripts removes them from "Addons currently running on this tab".

**To Reproduce**

1. Enable website dark mode
2. Open the "Addons" tab in the popup
3. Disable website dark mode
4. It's removed from the list of currently running addons, which means you have to scroll to the bottom to enable it again after closing and reopening the popup.

**Expected behavior**

It should stay under currently running addons until the tab is refreshed.

### PR DESCRIPTION
**Pull request by TheColaber**
Resolves #1303
Resolves #1670

Oh wow... #1500!

### CODE REVIEW

**Comment by RedGuy12** [comment]
Nice,  #1500!
________________________________

File: addons/load-extension/userscript.js [hunk 1]
```diff
@@ -0,0 +1,8 @@
+export default async function ({ addon, global, console }) {
+  // TODO: Add removing extensions.
```
Comment (RedGuy12) [inline addons/load-extension/userscript.js:2]: How could that work? The blocks would already be in the project...

File: addon-api/common/Settings.js [hunk 2]
```diff
@@ -21,6 +21,12 @@ export default class Settings extends Listenable {
     if (value === undefined) throw "ScratchAddons exception: invalid setting ID";
     else return value;
   }
+  /**
+   * @private
+   */
+  get getAll() {
+    return scratchAddons.globalState.addonSettings[this._addonId] || {};
```
Comment (RedGuy12) [inline addon-api/common/Settings.js:28]: the wiki will need to be updated

File: addons/load-extension/userscript.js (hunk 1 above)
Comment (TheColaber) [reply addons/load-extension/userscript.js:2]: For non-used

File: addon-api/common/Settings.js (hunk 2 above)
Comment (TheColaber) [reply addon-api/common/Settings.js:28]: yes

File: addons/load-extension/userscript.js (hunk 1 above)
Comment (RedGuy12) [reply addons/load-extension/userscript.js:2]: That's already something scratch does

Comment (TheColaber) [reply addons/load-extension/userscript.js:2]: no, you would have to reload the page for it to be removed. This addon should be able to allow it to be removed without refreshing.

Comment (RedGuy12) [reply addons/load-extension/userscript.js:2]: Scratch waits for the refresh because otherwise you would have time to take blocks out.

Comment (RedGuy12) [reply addons/load-extension/userscript.js:2]: Wait do you mean manually removing them? Probably a different addon would be best.

Comment (TheColaber) [reply addons/load-extension/userscript.js:2]: it fits the context enough. An addon that auto adds them, while also allowing removing.

Comment (mxmou) [reply addons/load-extension/userscript.js:2]: > it fits the context enough. An addon that auto adds them, while also allowing removing.

A single addon should not do multiple unrelated things.

File: addons/load-extension/userscript.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+  for (let ext of Object.keys(addon.settings.getAll)) {
```
Comment (WorldLanguages) [inline addons/load-extension/userscript.js:3]: Let's use normal `addon.settings.get()`

**Comment by WorldLanguages** [comment]
What am I missing, isn't adding an extension permanent? I don't think it's a good idea to have an addon that causes permanent changes to every project...

**Comment by TheColaber** [comment]
Adding an extension is not permanent. If not a single block from that extension is in your project, after a reload, it will go away.

**Review by WorldLanguages — APPROVED** [review]

**Comment by RedGuy12** [comment]
Did you completely change the topic of the pr lol

**Comment by TheColaber** [comment]
I had 2 things in one so W_L had me split it. Yes, sorta.

File: addons/addons.json [hunk 4]
```diff
@@ -72,5 +72,6 @@
   "comments-linebreaks",
   "account-settings-capitalize",
   "show-bbcode",
-  "copy-message-link"
+  "copy-message-link",
+  "load-extension"
```
Comment (RedGuy12) [inline addons/addons.json:76]: not needed

File: addon-api/common/Self.js [hunk 5]
```diff
@@ -0,0 +1,27 @@
+import Listenable from "./Listenable.js";
+
+/**
+ * @extends Listenable
+ */
+export default class Self extends Listenable {
+  constructor(that, info) {
+    super();
+    this.id = info.id;
+    this.extra = that;
```
Comment (WorldLanguages) [inline addon-api/common/Self.js:10]: Why?

File: addon-api/content-script/Tab.js [hunk 6]
```diff
@@ -26,6 +26,7 @@ export default class Tab extends Listenable {
     this.traps = new Trap(this);
     this.redux = new ReduxHandler();
     this._waitForElementSet = new WeakSet();
+    this.listeners = [];
```
Comment (WorldLanguages) [inline addon-api/content-script/Tab.js:29]: `_listeners`?

File: addons/editor-stage-left/fix-share-the-love.js [hunk 7]
```diff
@@ -1,27 +1,38 @@
 export default function ({ addon, global, console }) {
+  let interval, injected;
+
+  addon.self.addEventListener("addonDisabled", () => clearInterval(interval));
+  addon.self.addEventListener("addonReeabled", () => (injected ? 0 : tryInjecting()));
```
Comment (WorldLanguages) [inline addons/editor-stage-left/fix-share-the-love.js:5]: reeabled?

File: background/get-userscripts.js [hunk 8]
```diff
@@ -32,12 +32,11 @@ async function getContentScriptInfo(url) {
     addonsWithUserscripts: [],
     userstyleUrls: [],
     themes: [],
+    allAddons: [],
```
Comment (WorldLanguages) [inline background/get-userscripts.js:35]: What is this for?

File: addon-api/common/Self.js (hunk 5 above)
Comment (TheColaber) [reply addon-api/common/Self.js:10]: the "that" is needed below

File: background/handle-settings-page.js [hunk 9]
```diff
@@ -24,6 +24,13 @@ chrome.runtime.onMessage.addListener(function (request, sender, sendResponse) {
     chrome.storage.sync.set({
       addonsEnabled: scratchAddons.localState.addonsEnabled,
     });
+    chrome.tabs.query({}, (tabs) => {
+      // Tabs with urls means that ScratchAddons has perms there
+      tabs = tabs.filter((tab) => tab.url);
```
Comment (WorldLanguages) [inline background/handle-settings-page.js:29]: Keep in mind this is not the case in Firefox - not even Scratch URLs have a url property. For consistency I should check how I did this in other parts of background code and do it the same way, not sure

File: content-scripts/cs.js [hunk 10]
```diff
@@ -77,7 +82,11 @@ function injectUserstylesAndThemes({ userstyleUrls, themes, isUpdate }) {
       let css = theme.styles[styleUrl];
       // Replace %addon-self-dir% for relative URLs
       css = css.replace(/\%addon-self-dir\%/g, chrome.runtime.getURL(`addons/${theme.addonId}`));
-      if (isUpdate && css.startsWith("/* sa-autoupdate-theme-ignore */")) continue;
+      if (
+        isUpdate &&
+        (css.startsWith("/* sa-autoupdate-theme-ignore */") || css.startsWith("/* sa-theme-enable-ignore */"))
```
Comment (WorldLanguages) [inline content-scripts/cs.js:87]: What does `/* sa-theme-enable-ignore */"` do?

File: background/get-userscripts.js (hunk 8 above)
Comment (TheColaber) [reply background/get-userscripts.js:35]: "get-userscripts.js" needs the scripts that should be run for ALL addons. See where allAddons is used in cs.js and then in the other module.js file.

File: background/handle-settings-page.js (hunk 9 above)
Comment (TheColaber) [reply background/handle-settings-page.js:29]: We use this for the settings page I believe...

File: content-scripts/inject/run-userscript.js [hunk 11]
```diff
@@ -1,7 +1,10 @@
 import Addon from "../../addon-api/content-script/Addon.js";
+const ranAddons = [];
```
Comment (WorldLanguages) [inline content-scripts/inject/run-userscript.js:2]: This logic should likely not be in this file, which simply exports a function

File: content-scripts/inject/run-userscript.js [hunk 12]
```diff
@@ -1,7 +1,10 @@
...
 
-export default async function runAddonUserscripts({ addonId, scripts, traps }) {
-  const addonObj = new Addon({ id: addonId, traps });
+export default async function runAddonUserscripts({ addonId, scripts, traps }, { late = false } = {}) {
```
Comment (WorldLanguages) [inline content-scripts/inject/run-userscript.js:4]: What does "late" mean?

File: content-scripts/cs.js (hunk 10 above)
Comment (TheColaber) [reply content-scripts/cs.js:87]: Shouldn't enable userscript when addon enabled if that comment it there

File: content-scripts/inject/run-userscript.js (hunk 11 above)
Comment (TheColaber) [reply content-scripts/inject/run-userscript.js:2]: I think it needs to...

File: content-scripts/inject/run-userscript.js (hunk 12 above)
Comment (TheColaber) [reply content-scripts/inject/run-userscript.js:4]: It wasn't enabled on document start, thereby, late.

File: addon-api/common/Self.js (hunk 5 above)
Comment (WorldLanguages) [reply addon-api/common/Self.js:10]: Oh, name it `_addonObj`

File: content-scripts/cs.js [hunk 13]
```diff
@@ -50,6 +50,11 @@ chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
     sendResponse(initialUrl);
   } else if (request.themesUpdated) {
     injectUserstylesAndThemes({ themes: request.themesUpdated, isUpdate: true });
+  } else if (request.newAddonState) {
+    let addonId = request.newAddonState.addonId;
+    let obj = {};
+    obj[`saAddon${request.newAddonState.newState ? "En" : "Dis"}abled`] = addonId;
+    postMessage(obj, "*");
```
Comment (TheColaber) [inline content-scripts/cs.js:57]: Make this neater

File: addon-api/content-script/Addon.js [hunk 14]
```diff
@@ -9,8 +9,28 @@ import Tab from "./Tab.js";
 export default class UserscriptAddon extends Addon {
   constructor(info) {
     super(info);
+    this._addonId = info.id;
     this.__path = document.getElementById("scratch-addons").getAttribute("data-path");
     this.tab = new Tab(info);
+    this.self.disabled = false;
+    this.self.late = info.late;
```
Comment (WorldLanguages) [inline addon-api/content-script/Addon.js:16]: References to `late` should be removed from this PR, userscript autorefresh should be another PR.

File: addon-api/content-script/Addon.js [hunk 15]
```diff
@@ -9,8 +9,28 @@ import Tab from "./Tab.js";
...
+
+    addEventListener("message", (event) => {
```
Comment (WorldLanguages) [inline addon-api/content-script/Addon.js:18]: Use messaging through `template#scratch-addons`

File: background/get-userscripts.js (hunk 8 above)
Comment (WorldLanguages) [inline background/get-userscripts.js:35]: There's no need for `allAddons` unless you're implementing userscript autorefresh.

File: background/handle-settings-page.js (hunk 9 above)
Comment (WorldLanguages) [reply background/handle-settings-page.js:29]: TODO: check

File: content-scripts/cs.js (hunk 10 above)
Comment (WorldLanguages) [reply content-scripts/cs.js:86]: I don't think I understood what it does, can we have a proper written explanation?

**Comment by WorldLanguages** [comment]
TODO:
1. How to handle styles? Stuff gets sketchy specially in themes, that have their own rules on style autoupdating
2. Make sure `editor-stage-left` works properly with dynamic disabling, or put that into a separate PR.

**Comment by WorldLanguages** [comment]
I moved discussion about themes to #1672

**Comment by TheColaber** [comment]
Awaiting ScratchAddons/manifest-schema#26 for approval to merge.

**Review by CuriousG2 — APPROVED** [review]

**Review by WorldLanguages — APPROVED** [review]
